File: leetcode/265-paint-house-ii/main.go
package main

import (
	"fmt"
	"math"
)

/*
	intuitive approach: bfs
	- for each node, there are 2 options to add in the next level,
		therefore we can calculate all paths and get the min path sum

	Time	O(2^n)
	Space	O(1)
	TLE
*/

/*
	1st approach: dynamic programming
	- reuse the approach in I):leetcode 256

	mind the corner cases:
	e.g.1 []
	e.g.2 [[8]]
	e.g.3 [[8],[9]]

	Time	O(n)
	Space	O(1)
	4 ms, faster than 100.00%
*/
func minCost(costs [][]int) int {
	// cost0 := 0
	// cost1 := 0
	// cost2 := 0
	// for i := 0; i < len(costs); i++ {
	// 	// since we need to use cost<k> values, we need to store the next results with extra space
	// 	temp0 := costs[i][0] + findMin(cost1, cost2)
	// 	temp1 := costs[i][1] + findMin(cost0, cost2)
	// 	temp2 := costs[i][2] + findMin(cost0, cost1)
	// 	// set now
	// 	cost0 = temp0
	// 	cost1 = temp1
	// 	cost2 = temp2
	// }
	// return findMin(cost0, findMin(cost1, cost2))
	if len(costs) == 0 || len(costs[0]) == 0 {
		return 0
	}
	if len(costs[0]) == 1 {
		return costs[0][0]
	}
	dp := []int{}
	for i := 0; i < len(costs[0]); i++ {
		dp = append(dp, 0)
	}

	for i := 0; i < len(costs); i++ {
		temp := []int{}
		for j := 0; j < len(costs[i]); j++ {
			temp = append(temp, costs[i][j]+findMinExcept(dp, j))
		}
		dp = temp
	}
	res := findMinExcept(dp, -1)
	if res == math.MaxInt64 {
		return costs[0][0]
	}
	return res
}

func findMinExcept(nums []int, except int) int {
	res := math.MaxInt64
	for i, num := range nums {
		if i != except && num < res {
			res = num
		}
	}
	return res
}

func main() {
	// 10
	a := [][]int{
		{17, 2, 17},
		{16, 16, 5},
		{14, 3, 19},
	}
	fmt.Println(minCost(a))
	// 43
	a = [][]int{
		{5, 8, 6},
		{19, 14, 13},
		{7, 5, 12},
		{14, 15, 17},
		{3, 20, 10},
	}
	fmt.Println(minCost(a))
}

File: leetcode/265-paint-house-ii/main.py
import sys
from typing import List

"""
    1st: similar to lc256
    - but this time we can optimize the speed by precompute 2 smallest values from the previous row result

    Time    O(NK)
    Space   O(K)
    104 ms, faster than 95.10%
"""


class Solution:
    def minCostII(self, costs: List[List[int]]) -> int:
        if len(costs) == 0 or len(costs[0]) == 0:
            return 0
        R = len(costs)
        C = len(costs[0])
        dp = costs[0][:]
        for i in range(1, R):
            min1, min2 = sys.maxsize, sys.maxsize
            min1Idx, min2Idx = -1, -1
            for j in range(C):
                if dp[j] < min1:
                    min2, min2Idx = min1, min1Idx
                    min1, min1Idx = dp[j], j
                elif dp[j] < min2:
                    min2, min2Idx = dp[j], j
            temp = C * [0]
            for j in range(C):
                if j != min1Idx:
                    temp[j] = costs[i][j] + min1
                else:
                    temp[j] = costs[i][j] + min2
            dp = temp
        return min(dp)
